rfcs/0077-async-pipeline.md

```diff
@@ -0,0 +1,528 @@
+
+   A_local[0] <- A_shared[(i + 1) % 4, ...]
+   B_local[0] <- B_shared[(i + 1) % 4, ...]
+
+   compute(A_local[1], B_local[1])
+
+# Epilogue
+for i in range(3):
+   async_wait_stage(0, 1 - i)
+
+   A_local[1] <- A_shared[0, ...]
+   B_local[1] <- B_shared[0, ...]
+
+   compute(A_local[0], B_local[0])
+
+   if i < 2:
+      A_local[0] <- A_shared[0, ...]
+      B_local[0] <- B_shared[0, ...]
+
+   compute(A_local[1], B_local[1])
+
+```
+
+### Implicit vs explicit approach to synchronization
+
+The model of async synchronization adopted by CUDA can be categorized as an “implicit” one: Instead of saying “Wait for this operation to complete”, it says “Wait until only N most recent async operations are in flight”, or equivalently, “Wait until M oldest async operates have completed”, where M = “number of async operations in flight” - N.
+
+In contrast, a standard and intuitive approach is more explicit, e.g. wait for the operation associated with “this” token / future to complete etc. This is true for “async-await” in general-purpose languages, [Async](https://mlir.llvm.org/docs/Dialects/AsyncDialect/) and [NVGPU](https://mlir.llvm.org/docs/Dialects/NVGPU/) dialects in MLIR, for example.
+
+In general, the explicit approach is probably more preferable, since
+
+- It makes it obvious which operation is waiting on which
+- It is less stateful (less assumption on how the underlying HW should work)
+- It naturally handles synchronization in the presence of control flow (since we can only wait on an operation that has actually happened).
+
+These properties may help if we want do some analysis of async programs.
+
+The current design started from and has stayed with CUDA’s implicit synchronization model based on counting, primarily because it makes mapping to the corresponding PTX instructions trivial. We can adopt the explicit model instead, if we have a good way to translate token-based synchronization to the counting one for PTX. So far, we do not have a good solution for this. MLIR has adopted the token abstraction, but they have not solve this problem either: Their `DeviceAsyncWaitOp` has [an optional attribute `numGroups`](https://mlir.llvm.org/docs/Dialects/NVGPU/#nvgpudevice_async_wait-mlirnvgpudeviceasyncwaitop) that directly corresponds to "in-flight count", and they basically generate either `wait(numGroups)` or `wait(0)`, [in their translation](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp#L426-L427) of  `DeviceAsyncWaitOp` (token based) to PTX `cp.async` (counting based). `wait(0)` is always correct but least precise / efficient.
+
+
+
+(The following is highly speculative) On the other hand, translation from “count” to “token” seems more feasible: At each synchronization point, a backend presumably maintains the number and the order of pending async operations. Given the count, it should be possible to derive the correct token from the corresponding ordered list of tokens.
```

Comment:
@masahi I believe I understand the intended usage of "stage" here, but _could_ we use it like Eric is suggesting? Is there a limitation to having an arbitrary number of async stages?

@Lunderberg Unless I've misunderstood, this translation would require fully unrolling all loops, right? Because each time you call `commit` you need to pass in a new static integer?

Reply:
@masahi is accurate. In TIR software pipeline, we can group multiple statements into one stage. Here stage is only an annotation of how the loop should be shifted, the i-th iteration of the pipelined loop executes `i+stage`-th iteration in the original loop

Reply:
> @masahi I believe I understand the intended usage of "stage" here, but could we use it like Eric is suggesting? Is there a limitation to having an arbitrary number of async stages?

That's correct.  My goal was to map out what can be expressed using the commit/wait semantics defined, whether it results in broader functionality than the pipelinining use case, and whether that broader functionality is desired.

> On the other hand, I and this proposal talk about "stage" in the sense used by TIR software pipeline. So for example, if I am given an annotation "software_pipeline_stage", ann_val=[0, 0, 3], I'd say there are "two" stages in the TIR sense, even though the traditional sense would be four stages.

Reading through again, I think there are two different levels of abstraction, and that the use of the term "stage" in both levels may be causing my confusion.  At the higher abstraction level with `"software_pipeline_stage"`, each stage is defined by their values in the annotations.  At the lower abstraction level with commit/wait, the first argument defines a a set of work to be completed, or to be waited on.  The first argument at the commit/wait abstraction level is generated from the stage at the `"software_pipeline"` abstraction level, but that doesn't mean it would be the only possible lowering.

> @Lunderberg Unless I've misunderstood, this translation would require fully unrolling all loops, right? Because each time you call commit you need to pass in a new static integer?
>
> An alternative to unrolling is if non-constant stage numbers are allowed. It might not be pretty or succinct, but you could thread variable(s) through the program to hold stage (fence) IDs.

@JosephTheOctonaut Correct, the non-constant stage numbers are what I had been picturing.  For static graphs without unrolling loops, this would be some `loop_iter + offset` used to define the unique ID for each commit/wait.